demux/demux_lavf.c

```diff
@@ -551,6 +562,25 @@ static void select_tracks(struct demuxer *demuxer, int start)
         bool selected = stream && demux_stream_is_selected(stream) &&
                         !stream->attached_picture;
         st->discard = selected ? AVDISCARD_DEFAULT : AVDISCARD_ALL;
+
+        if (!stream)
+            continue;
+        if (stream->type == STREAM_VIDEO) {
+            if (selected) {
+                priv->program.num = stream->program_num;
+                priv->selected_video_stream = stream;
+            } else if (priv->program.num &&
+                       priv->program.num == stream->program_num) {
+                // always demux all video streams within the same program
+                st->discard = AVDISCARD_DEFAULT;
+            }
+        }
+        if (stream->type == STREAM_AUDIO) {
+            if (selected)
+                priv->selected_audio_stream = stream;
+            if (stream == priv->active_audio_stream)
+                st->discard = AVDISCARD_DEFAULT;
```

Comment:
Wouldn't it be nicer to set `st->discard` in a single place, e.g. setting the `selected` variable instead?